test/e2e/dynamic_resource_e2e_test.go

```diff
@@ -1,105 +1,163 @@
 
-		c := newKubeClient(GinkgoT())
-		crc := newCRClient(GinkgoT())
-		dynamicClient := ctx.Ctx().DynamicClient()
+	var (
+		c             operatorclient.ClientInterface
+		crc           versioned.Interface
+		dynamicClient dynamic.Interface
+	)
 
-		ns, err := c.KubernetesInterface().CoreV1().Namespaces().Create(&corev1.Namespace{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: genName("ns-"),
-			},
-		})
-		require.NoError(GinkgoT(), err)
-
-		deleteOpts := &metav1.DeleteOptions{}
-		defer func() {
-			require.NoError(GinkgoT(), c.KubernetesInterface().CoreV1().Namespaces().Delete(ns.GetName(), deleteOpts))
-		}()
-
-		catsrc := &v1alpha1.CatalogSource{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      genName("dynamic-catalog-"),
-				Namespace: ns.GetName(),
-			},
-			Spec: v1alpha1.CatalogSourceSpec{
-				Image:      "quay.io/olmtest/catsrc_dynamic_resources:e2e-test",
-				SourceType: v1alpha1.SourceTypeGrpc,
-			},
-		}
-
-		catsrc, err = crc.OperatorsV1alpha1().CatalogSources(catsrc.GetNamespace()).Create(catsrc)
-		require.NoError(GinkgoT(), err)
-
-		defer func() {
-			require.NoError(GinkgoT(), crc.OperatorsV1alpha1().CatalogSources(catsrc.GetNamespace()).Delete(catsrc.GetName(), deleteOpts))
-		}()
-
-		// Wait for the CatalogSource to be ready
-		catsrc, err = fetchCatalogSource(GinkgoT(), crc, catsrc.GetName(), catsrc.GetNamespace(), catalogSourceRegistryPodSynced)
-		require.NoError(GinkgoT(), err)
-
-		// Generate a Subscription
-		subName := genName("dynamic-resources")
-		cleanupSub := createSubscriptionForCatalog(GinkgoT(), crc, catsrc.GetNamespace(), subName, catsrc.GetName(), "etcd", "singlenamespace-alpha", "", v1alpha1.ApprovalAutomatic)
-		defer cleanupSub()
-
-		sub, err := fetchSubscription(GinkgoT(), crc, catsrc.GetNamespace(), subName, subscriptionHasInstallPlanChecker)
-		require.NoError(GinkgoT(), err)
-
-		// Wait for the expected InstallPlan's execution to either fail or succeed
-		ipName := sub.Status.InstallPlanRef.Name
-		ip, err := waitForInstallPlan(GinkgoT(), crc, ipName, sub.GetNamespace(), buildInstallPlanPhaseCheckFunc(v1alpha1.InstallPlanPhaseFailed, v1alpha1.InstallPlanPhaseComplete))
-		require.NoError(GinkgoT(), err)
-		// Ensure the InstallPlan contains the steps resolved from the bundle image
-		expectedSteps := map[registry.ResourceKey]struct{}{
-			{Name: "my-prometheusrule", Kind: "PrometheusRule"}: {},
-			{Name: "my-servicemonitor", Kind: "ServiceMonitor"}: {},
-		}
-
-		for _, step := range ip.Status.Plan {
-			key := registry.ResourceKey{
-				Name: step.Resource.Name,
-				Kind: step.Resource.Kind,
-			}
-			for expected := range expectedSteps {
-				if strings.HasPrefix(key.Name, expected.Name) && key.Kind == expected.Kind {
-					delete(expectedSteps, expected)
-				}
+	BeforeEach(func() {
+		c = newKubeClient()
+		crc = newCRClient()
+		dynamicClient = ctx.Ctx().DynamicClient()
+	})
+
+	Context("Create namespace and CatalogSource ", func() {
+		var (
+			ns         *corev1.Namespace
```

Comment:
The `ns` variable's type is supposed to be `*v1.Namespace` as per the Create() method's return type.
`Create(*v1.Namespace) (*v1.Namespace, error)`
 But the compiler kept complaining, but for some reason, it worked when I changed the type to `*corev1.Namespace`
Not sure if it is correct?

Reply:
That's correct, check out line 10:

```go
corev1 "k8s.io/api/core/v1"
```
`Create`'s signature is written in terms of the package it comes from, so the docs for it just say `v1`, but when that package is imported into this file, we rename it to `corev1`. The compiler will check that they refer to the same type

Reply:
Oh, my bad. I forgot to notice that the package was using an alias.